File: core/src/main/scala/kafka/network/Channel.scala
/**
  * Licensed to the Apache Software Foundation (ASF) under one or more
  * contributor license agreements.  See the NOTICE file distributed with
  * this work for additional information regarding copyright ownership.
  * The ASF licenses this file to You under the Apache License, Version 2.0
  * (the "License"); you may not use this file except in compliance with
  * the License.  You may obtain a copy of the License at
  *
  *    http://www.apache.org/licenses/LICENSE-2.0
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
  * See the License for the specific language governing permissions and
  * limitations under the License.
  */

package kafka.network

import java.io.IOException
import java.nio.ByteBuffer
import java.nio.channels.ReadableByteChannel
import java.nio.channels.GatheringByteChannel
import java.nio.channels.SelectionKey
import java.nio.channels.Selector
import java.nio.channels.SocketChannel
import java.nio.channels.ClosedChannelException
import java.nio.channels.spi.AbstractSelectableChannel

import kafka.utils.Logging

class Channel( var socketChannel: SocketChannel ) extends ReadableByteChannel with GatheringByteChannel with Logging {

  /**
    * Returns true if the network buffer has been flushed outu and is empty.
    */
  def flush() : Boolean = {
    true
  }

  /**
    * Closes this channel.
    *
    * @throws IOException If an I/O error occurs
    */
  @throws(classOf[IOException])
  def close() {
    socketChannel.socket().close()
    socketChannel.close()
  }

  @throws(classOf[IOException])
  def close(force: Boolean) {
    if(isOpen || force) close()
  }

  /**
   * Tells wheter or not this channel is open.
   */
  override def isOpen: Boolean = {
    socketChannel.isOpen()
  }

  /**
   * Writes a sequence of bytes to this channel from the given buffer.
   */
  @throws(classOf[IOException])
  override def write(src: ByteBuffer) : Int = {
    socketChannel.write(src)
  }

  @throws(classOf[IOException])
  override def write(srcs: Array[ByteBuffer]) : Long = {
    socketChannel.write(srcs)
  }

  @throws(classOf[IOException])
  override def write(srcs: Array[ByteBuffer], offset: Int, length: Int) : Long = {
    socketChannel.write(srcs, offset, length)
  }

  @throws(classOf[IOException])
  override def read(dst: ByteBuffer) : Int = {
    socketChannel.read(dst)
  }

  def getIOChannel : SocketChannel = {
    socketChannel
  }

  def isHandshakeComplete(): Boolean = {
    true
  }

  @throws(classOf[ClosedChannelException])
  def register(selector: Selector, ops: Int): SelectionKey = {
    socketChannel.register(selector, ops)
  }

  @throws(classOf[IOException])
  def flushOutbound: Boolean = {
    false
  }


  // /**
  //   * Selector related methods will be delegated to the underlying channel
  //   */
  // override def blockingLock: Object = {
  //   socketChannel.blockingLock()
  // }

  // override def configureBlocking(block: Boolean): SelectableChannel = {
  //   socketChannel.configureBlocking(block)
  // }

  // override def implCloseChannel() {
  //   socketChannel.implCloseChannel()
  // }

  /**
    * Performs SSL handshake hence is a no-op for the non-secure
    * implementation
    * @param read Unused in non-secure implementation
    * @param write Unused in non-secure implementation
    * @return Always return 0
    * @throws IOException
    */
  @throws(classOf[IOException])
  def handshake(read: Boolean, write: Boolean): Int = {
    return 0
  }

  override def toString(): String = {
    super.toString()+":"+this.socketChannel.toString()
  }

  def getOutboundRemaining(): Int = {
    return 0
  }
}

File: clients/src/main/java/org/apache/kafka/common/network/Channel.java
/**
 * Licensed to the Apache Software Foundation (ASF) under one or more
 * contributor license agreements.  See the NOTICE file distributed with
 * this work for additional information regarding copyright ownership.
 * The ASF licenses this file to You under the Apache License, Version 2.0
 * (the "License"); you may not use this file except in compliance with
 * the License.  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package org.apache.kafka.common.network;


import java.io.IOException;

import java.net.Socket;
import java.nio.channels.SelectionKey;

import java.security.Principal;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;


public class Channel {
    private static final Logger log = LoggerFactory.getLogger(Channel.class);
    private final int id;
    private TransportLayer transportLayer;
    private Authenticator authenticator;
    private NetworkReceive receive;
    private NetworkSend send;

    public Channel(int id, TransportLayer transportLayer, Authenticator authenticator) throws IOException {
        this.id = id;
        this.transportLayer = transportLayer;
        this.authenticator = authenticator;
    }

    public void close() throws IOException {
        transportLayer.close();
        authenticator.close();
    }

    /**
     * returns user principal for the session
     * In case of PLAINTEXT and No Authentication returns ANONYMOUS as the userPrincipal
     * If SSL used without any SASL Authentication returns SSLSession.peerPrincipal
     */
    public Principal principal() throws IOException {
        return authenticator.principal();
    }

    /**
     * Does handshake of transportLayer and Authentication using configured authenticator
     */
    public void prepare() throws IOException {
        if (transportLayer.isReady() && authenticator.isComplete())
            return;
        if (!transportLayer.isReady())
            transportLayer.handshake();
        if (transportLayer.isReady() && !authenticator.isComplete())
            authenticator.authenticate();
    }

    public void disconnect() {
        transportLayer.disconnect();
    }


    public void finishConnect() throws IOException {
        transportLayer.finishConnect();
    }

    public int id() {
        return id;
    }

    public void mute() {
        transportLayer.removeInterestOps(SelectionKey.OP_READ);
    }

    public void unmute() {
        transportLayer.addInterestOps(SelectionKey.OP_READ);
    }

    public boolean isReady() {
        return transportLayer.isReady() && authenticator.isComplete();
    }

    public String socketDescription() {
        Socket socket = transportLayer.socketChannel().socket();
        if (socket == null)
            return "[unconnected socket]";
        else if (socket.getInetAddress() != null)
            return socket.getInetAddress().toString();
        else
            return socket.getLocalAddress().toString();
    }

    public void setSend(NetworkSend send) {
        if (this.send != null)
            throw new IllegalStateException("Attempt to begin a send operation with prior send operation still in progress.");
        this.send = send;
        this.transportLayer.addInterestOps(SelectionKey.OP_WRITE);
    }

    public NetworkReceive read() throws IOException {
        NetworkReceive result = null;

        if (receive == null) {
            receive = new NetworkReceive(id);
        }
        receive(receive);
        if (receive.complete()) {
            receive.payload().rewind();
            result = receive;
            receive = null;
        }
        return result;
    }

    public NetworkSend write() throws IOException {
        NetworkSend result = null;
        if (send != null && send(send)) {
            result = send;
            send = null;
        }
        return result;
    }

    private long receive(NetworkReceive receive) throws IOException {
        long result = receive.readFrom(transportLayer);
        return result;
    }

    private boolean send(NetworkSend send) throws IOException {
        send.writeTo(transportLayer);
        if (send.completed()) {
            transportLayer.removeInterestOps(SelectionKey.OP_WRITE);
        }
        return send.completed();
    }

}
